# Document a California Housing Price Prediction regression model

Use the [California Housing Price Prediction](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_california_housing.html) sample dataset from Sklearn to train a simple regression model and document that model with the ValidMind Developer Framework.

As part of the notebook, you will learn how to train a sample model while exploring how the documentation process works:

- Initializing the ValidMind Developer Framework
- Loading a sample dataset provided by the library to train a simple regression model
- Running a ValidMind test suite to quickly generate documention about the data and model


## About ValidMind

ValidMind's platform enables organizations to identify, document, and manage model risks for all types of models, including AI/ML models, LLMs, and statistical models. As a model developer, you use the ValidMind Developer Framework to automate documentation and validation tests, and then use the ValidMind AI Risk Platform UI to collaborate on model documentation. Together, these products simplify model risk management, facilitate compliance with regulations and institutional standards, and enhance collaboration between yourself and model validators.

If this is your first time trying out ValidMind, you can make use of the following resources alongside this notebook:

- [Get started](https://docs.validmind.ai/get-started/get-started.html) — The basics, including key concepts, and how our products work
- [Get started with the ValidMind Developer Framework](https://docs.validmind.ai/developer/get-started-developer-framework.html) — The path for developers, more code samples, and our developer reference


## Before you begin

<div class="alert alert-block alert-info" style="background-color: #f7e4ee; color: black; border: 1px solid black;"><b>For access to all features available in this notebook, create a free ValidMind account.</b>
<br></br>
Signing up is FREE — <a href="https://docs.validmind.ai/guide/configuration/register-with-validmind.html"><b>Register with ValidMind</b></a></div>

If you encounter errors due to missing modules in your Python environment, install the modules with `pip install`, and then re-run the notebook. For more help, refer to [Installing Python Modules](https://docs.python.org/3/installing/index.html).


## Install the client library

The client library provides Python support for the ValidMind Developer Framework. To install it:


In [1]:
%pip install -q validmind


[notice] A new release of pip is available: 23.3.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


## Initialize the client library

ValidMind generates a unique _code snippet_ for each registered model to connect with your developer environment. You initialize the client library with this code snippet, which ensures that your documentation and tests are uploaded to the correct model when you run the notebook.

Get your code snippet:

1. In a browser, [log in to ValidMind](https://docs.validmind.ai/guide/configuration/log-in-to-validmind.html).

2. In the left sidebar, navigate to **Model Inventory** and click **+ Register new model**.

3. Enter the model details, making sure to select **Binary classification** as the template and **Marketing/Sales - Attrition/Churn Management** as the use case, and click **Continue**. ([Need more help?](https://docs.validmind.ai/guide/model-inventory/register-models-in-inventory.html))

4. Go to **Getting Started** and click **Copy snippet to clipboard**.

Next, replace this placeholder with your own code snippet:


In [2]:
# Replace with your code snippet

import validmind as vm


import validmind as vm

vm.init(
  api_host = "...",
  api_key = "...",
  api_secret = "...",
  project = "..."
)


import logging
import sys
import site

logger = logging.getLogger()

# Print the path to the current Python interpreter
logger.info("Python executable path: " + sys.executable)

# Print the path to the site-packages directory
logger.info("Site-packages path: " + str(site.getsitepackages()))


2024-10-01 14:12:38,343 - INFO(validmind.api_client): 🎉 Connected to ValidMind!
📊 Model: [Int. Tests] Regression (ID: cltnl7swd000s1omg6t4ul5bn)
📁 Document Type: model_documentation


## Initialize the Python environment

Next, let's import the necessary libraries and set up your Python environment for data analysis:


In [3]:
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

%matplotlib inline

### Preview the documentation template

A template predefines sections for your model documentation and provides a general outline to follow, making the documentation process much easier.

You will upload documentation and test results into this template later on. For now, take a look at the structure that the template provides with the `vm.preview_template()` function from the ValidMind library and note the empty sections:


In [4]:
vm.preview_template()

2024-10-01 14:12:38,920 - ERROR(validmind.tests.load): Unable to load test validmind.model_validation.sklearn.RegressionModelsPerformanceComparison. No module named 'validmind.tests.model_validation.sklearn.RegressionModelsPerformanceComparison'


Accordion(children=(Accordion(children=(HTML(value='<p>Empty Section</p>'), Accordion(children=(HTML(value='<p…

## Load the sample dataset

To be able to use a sample dataset, you need to import the dataset and load it into a pandas [DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html), a two-dimensional tabular data structure that makes use of rows and columns:


## Load the sample dataset

The sample dataset used here is provided by the ValidMind library. To be able to use it, you need to import the dataset and load it into a pandas [DataFrame](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html), a two-dimensional tabular data structure that makes use of rows and columns:


In [5]:
# Import the sample dataset from the library

from validmind.datasets.regression import california_housing as demo_dataset

print(f"Loaded demo dataset with: \n\n\t• Target column: '{demo_dataset.target_column}")

raw_df = demo_dataset.load_data()
raw_df.head()

Loaded demo dataset with: 

	• Target column: 'MedHouseVal


HouseAge  AveRooms  AveBedrms  Population  AveOccup  MedHouseVal
0      41.0  6.984127   1.023810       322.0  2.555556        4.526
1      21.0  6.238137   0.971880      2401.0  2.109842        3.585
2      52.0  8.288136   1.073446       496.0  2.802260        3.521
3      52.0  5.817352   1.073059       558.0  2.547945        3.413
4      52.0  6.281853   1.081081       565.0  2.181467        3.422

## Document the model

As part of documenting the model with the ValidMind Developer Framework, you need to preprocess the raw dataset, initialize some training and test datasets, initialize a model object you can use for testing, and then run the full suite of tests.


### Prepocess the raw dataset

Preprocessing performs a number of operations to get ready for the subsequent steps:

- Preprocess the data: Splits the DataFrame (`df`) into multiple datasets (`train_df`, `validation_df`, and `test_df`) using `demo_dataset.preprocess` to simplify preprocessing.
- Separate features and targets: Drops the target column to create feature sets (`x_train`, `x_val`) and target sets (`y_train`, `y_val`).
- Initialize RandomForestRegressor regressor: Creates an `RandomForestRegressor` object with random state set to 0.
- Set evaluation metrics: Specifies metrics for model evaluation as "errors" and "r2".
- Fit the model: Trains the model on `x_train` and `y_train` using the validation set `(x_val, y_val)`. Verbose output is disabled.


In [6]:
train_df, validation_df, test_df = demo_dataset.preprocess(raw_df)

x_train = train_df.drop(demo_dataset.target_column, axis=1)
y_train = train_df[demo_dataset.target_column]
x_val = validation_df.drop(demo_dataset.target_column, axis=1)
y_val = validation_df[demo_dataset.target_column]

Here we create two regression models so that the performance of the model can be compared through ValidMind test suite.


In [7]:
scale = False
if scale:
    scaler = StandardScaler()
    x_train = scaler.fit_transform(x_train)
    x_val = scaler.fit_transform(x_val)

model = RandomForestRegressor(random_state=0)
model.fit(x_train, y_train)
s1 = model.score(x_train, y_train)
s2 = model.score(x_val, y_val)
print("R² of Support Vector Regressor on training set: {:.3f}".format(s1))
print("R² of Support Vector Regressor on test set: {:.3f}".format(s2))

model_1 = GradientBoostingRegressor(random_state=0, max_depth=4)
model_1.fit(x_train, y_train)
model1_s1 = model_1.score(x_train, y_train)
model1_s2 = model_1.score(x_val, y_val)
print(
    "R² of Support Gradient Boosting Regressor on training set: {:.3f}".format(
        model1_s1
    )
)
print("R² of Support Gradient Boosting Regressor on test set: {:.3f}".format(model1_s2))

R² of Support Vector Regressor on training set: 0.918
R² of Support Vector Regressor on test set: 0.407
R² of Support Gradient Boosting Regressor on training set: 0.508
R² of Support Gradient Boosting Regressor on test set: 0.439


### Initialize the ValidMind datasets

Before you can run tests, you must first initialize a ValidMind dataset object using the [`init_dataset`](https://docs.validmind.ai/validmind/validmind.html#init_dataset) function from the ValidMind (`vm`) module.

This function takes a number of arguments:

- `dataset` — the raw dataset that you want to provide as input to tests
- `input_id` - a unique identifier that allows tracking what inputs are used when running each individual test
- `target_column` — a required argument if tests require access to true values. This is the name of the target column in the dataset

With all datasets ready, you can now initialize the raw, training and test datasets (`raw_df`, `train_df` and `test_df`) created earlier into their own dataset objects using [`vm.init_dataset()`](https://docs.validmind.ai/validmind/validmind.html#init_dataset):


In [8]:
vm_raw_dataset = vm.init_dataset(
    dataset=raw_df,
    input_id="raw_dataset",
    target_column=demo_dataset.target_column,
)

vm_train_ds = vm.init_dataset(
    dataset=train_df, input_id="train_dataset", target_column=demo_dataset.target_column
)

vm_test_ds = vm.init_dataset(
    dataset=test_df, input_id="test_dataset", target_column=demo_dataset.target_column
)

2024-10-01 14:12:45,195 - INFO(validmind.client): Pandas dataset detected. Initializing VM Dataset instance...
2024-10-01 14:12:46,260 - INFO(validmind.client): Pandas dataset detected. Initializing VM Dataset instance...
2024-10-01 14:12:46,460 - INFO(validmind.client): Pandas dataset detected. Initializing VM Dataset instance...


### Initialize a model object

Additionally, you need to initialize a ValidMind model objects (`vm_model` and `vm_model_1`) that can be passed to other functions for analysis and tests on the data. You simply intialize this model object with [`vm.init_model()`](https://docs.validmind.ai/validmind/validmind.html#init_model):


In [9]:
vm_model = vm.init_model(
    model,
    input_id="random_forest_regressor",
)
vm_model_1 = vm.init_model(
    model_1,
    input_id="gradient_boosting_regressor",
)

### Assign predictions to the datasets

We can now use the assign_predictions() method from the Dataset object to link existing predictions to any model. If no prediction values are passed, the method will compute predictions automatically:


In [10]:
vm_train_ds.assign_predictions(
    model=vm_model,
)
vm_train_ds.assign_predictions(
    model=vm_model_1,
)
vm_test_ds.assign_predictions(
    model=vm_model,
)
vm_test_ds.assign_predictions(
    model=vm_model_1,
)

2024-10-01 14:12:47,856 - INFO(validmind.vm_models.dataset.utils): Running predict_proba()... This may take a while
2024-10-01 14:12:47,856 - INFO(validmind.vm_models.dataset.utils): Not running predict_proba() for unsupported models.
2024-10-01 14:12:47,857 - INFO(validmind.vm_models.dataset.utils): Running predict()... This may take a while
/Users/anilsorathiya/Library/Caches/pypoetry/virtualenvs/validmind-1QuffXMV-py3.10/lib/python3.10/site-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(
2024-10-01 14:12:48,055 - INFO(validmind.vm_models.dataset.utils): Done running predict()
2024-10-01 14:12:48,056 - INFO(validmind.vm_models.dataset.dataset): No probabilities computed or provided. Not adding probability column to the dataset.
2024-10-01 14:12:48,057 - INFO(validmind.vm_models.dataset.utils): Running predict_proba()... This may take a while
2024-10-01 14:12:48,057 - INFO(validmin

### Run the full suite of tests

This is where it all comes together: you are now ready to run the documentation tests for the model as defined by the documentation template you looked at earlier.

The [`vm.run_documentation_tests`](https://docs.validmind.ai/validmind/validmind.html#run_test_suite) function finds and runs every tests specified in the test suites and then uploads all the documentation and test artifacts that get generated to the ValidMind AI Risk Platform.

The function takes two arguments:

- `dataset`: The data to be tested, specified as `vm_dataset`.
- `model`: The candidate model to be used for testing, specified as `vm_model`. -`models`: The list of models that can be compare with candidate model.

The variable `full_suite` then holds the result of these tests.


In [11]:
full_suite = vm.run_documentation_tests(
    inputs={
        "dataset": vm_train_ds,
        "datasets": (vm_train_ds, vm_test_ds),
        "model": vm_model,
        "models":[vm_model_1]
    }
)

2024-10-01 14:12:48,156 - ERROR(validmind.tests.load): Unable to load test validmind.model_validation.sklearn.RegressionModelsPerformanceComparison. No module named 'validmind.tests.model_validation.sklearn.RegressionModelsPerformanceComparison'


## Next steps

You can look at the results of this test suite right in the notebook where you ran the code, as you would expect. But there is a better way: view the test results as part of your model documentation right in the ValidMind Platform UI:

1. In the ValidMind Platform UI, go to the **Documentation** page for the model you registered earlier. ([Need more help?](https://docs.validmind.ai/guide/model-documentation/working-with-model-documentation.html)

2. Expand the following sections and take a look around:

   - **2. Data Preparation**
   - **3. Model Development**

What you can see now is a much more easily consumable version of the documentation, including the results of the tests you just performed, along with other parts of your model documentation that still need to be completed. There is a wealth of information that gets uploaded when you run the full test suite, so take a closer look around, especially at test results that might need attention (hint: some of the tests in **2.1 Data description** look like they need some attention).

If you want to learn more about where you are in the model documentation process, take a look at [Get started with the ValidMind Developer Framework](https://docs.validmind.ai/developer/get-started-developer-framework.html).
